velox/docs/develop/types.rst

```diff
@@ -2,54 +2,107 @@
+Velox supports both scalar types and complex types.
+Scalar types are categorized into a fixed set of physical types,
+and an extensible set of logical types.
+Physical types determine the in-memory layout of the data.
+Logical types add additional semantics to a physical type.
 
-Scalar Types
-~~~~~~~~~~~~
-Scalar types in Velox are logical and SQL-compatible.
-Each scalar type is implemented using a C++ type. The table
-below shows the supported scalar types and their corresponding C++ type.
-
-======================  ===========================    ==================
-Velox Type              C++ Type                       Bytes per Value
-======================  ===========================    ==================
-BOOLEAN                 bool                            0.125 (i.e. 1 bit)
-TINYINT                 int8_t                          1
-SMALLINT                int16_t                         2
-INTEGER                 int32_t	                        4
-BIGINT                  int64_t                         8
-DATE                    struct Date                     8
-REAL                    float                           4
-DOUBLE                  double                          8
-SHORT_DECIMAL           struct UnscaledShortDecimal     8
-LONG_DECIMAL            struct UnscaledLongDecimal     16
-TIMESTAMP               struct Timestamp               16
-INTERVAL DAY TO SECOND  struct IntervalDayTime          8
-VARCHAR                 struct StringView              16
-VARBINARY               struct StringView              16
-======================  ===========================    ==================
-
-All scalar types except DECIMAL have a one-to-one mapping with their C++ types.
-
-DECIMAL type is a special scalar type since it carries additional `precision`,
-and `scale` information. Similar to SQL decimal, `precision` is the number of
+Physical Types
+~~~~~~~~~~~~~~
+Each physical type is implemented using a C++ type. The table
+below shows the supported physical types, their corresponding C++ type,
+and bytes required per value.
+
+================   ===========================   ===================
+Physical Type      C++ Type                      Bytes per Value
+================   ===========================   ===================
+BOOLEAN            bool                          0.125 (i.e. 1 bit)
+TINYINT            int8_t                        1
+SMALLINT           int16_t                       2
+INTEGER            int32_t	                     4
+BIGINT             int64_t                       8
+HUGEINT            int128_t                      16
+DATE               struct Date                   8
+REAL               float                         4
+DOUBLE             double                        8
+TIMESTAMP          struct Timestamp              16
+VARCHAR            struct StringView             16
+VARBINARY          struct StringView             16
+OPAQUE             std::shared_ptr<void>         User-defined
```

Comment:
> User-defined

Hmm... VARCHAR says 16 though 16 bytes is only fixed-width part. OPAQUE also has fixed-width part that's not user-defined. 

Let's clarify further that some types use fixed number of bytes while other types use variable-width number of bytes per value. Variable-width values have fixed-width part stored in the 'values' buffer of a FlatVector<T> and variable-width part stored somewhere else. VARCHAR and VARBINARY types store variable width parts in 'stringBuffers' of the FlatVector<StringView>. OPAQUE types store variable width parts outside of the vector. Let's then rename this column to Fixed Width (bytes).

Reply:
Good point!